src/main/java/com/mraof/minestuck/computer/editmode/ClientEditToolDrag.java

```diff
@@ -0,0 +1,462 @@
+public class ClientEditToolDrag
+{
+	
+	@SubscribeEvent
+	public static void onClientTick(TickEvent.ClientTickEvent event)
+	{
+		if(event.side == LogicalSide.CLIENT)
+		{
+			ClientEditToolDrag.doRecycleCode(event);
+			ClientEditToolDrag.doReviseCode(event);
+		}
+	}
+	
+	@SubscribeEvent
+	public static void renderWorld(RenderLevelStageEvent event)
+	{
+		ClientEditToolDrag.renderOutlines(event);
+	}
+	
+	/**
+	 * Checks the conditions to see if the recycle/revise tool should be ended early.
+	 * @param isRevise If true, checks conditions for Revise, otherwise checks Recycle.
+	 * @param player Current client-side player.
+	 * @param toolKey Mouse key used to activate the given tool.
+	 * @return True if the tool should be canceled, false if it shouldn't.
+	 */
+	public static boolean shouldCancelDrag(boolean isRevise, Player player, KeyMapping toolKey)
+	{
+		if(toolKey.isDown())
+			if(isRevise ? !canEditRevise(player) : !canEditRecycle(player))
+				return true;
+		return false;
+	}
+	
+	/**
+	 * Resets the drag tool, and removes the server-cursor if the given edit tool is active.
+	 * @param isRevise If true, checks conditions for Revise, otherwise checks Recycle.
+	 * @param cap The current edit-tools capability.
+	 */
+	public static void cancelDrag(boolean isRevise, IEditTools cap)
+	{
+		if(cap.getToolMode() == (isRevise ? IEditTools.ToolMode.REVISE : IEditTools.ToolMode.RECYCLE))
+			MSPacketHandler.sendToServer(new EditmodeDragPacket.Reset());
+		cap.resetDragTools();
+	}
+	
+	/**
+	 * Determines whether or not the game should try to begin a drag tool.
+	 * @param toolKey The given tool's key.
+	 * @param isEditDragging Whether the game has started dragging already.
+	 * @return True if you're holding the key, but the game hasn't started a selection yet.
+	 */
+	public static boolean shouldBeginDrag(KeyMapping toolKey, boolean isEditDragging) { return toolKey.isDown() && !isEditDragging; }
+	
+	/**
+	 * Attempts to get the currently highlighted block.
+	 * If the player is highlighting a block, initialize the target edit-tool's parameters.
+	 * @param isRevise If true, uses Revise, otherwise uses Recycle.
+	 * @param cap The current edit-tools capability.
+	 * @param player Current client-side player.
+	 */
+	public static void tryBeginDrag(boolean isRevise, IEditTools cap, Player player)
+	{
+		BlockHitResult blockHit = getPlayerPOVHitResult(player.getLevel(), player);
+		if (blockHit.getType() == BlockHitResult.Type.BLOCK)
+		{
+			cap.beginDragTools(isRevise ? IEditTools.ToolMode.REVISE : IEditTools.ToolMode.RECYCLE, blockHit, player);
+		}
+	}
+	
+	/**
+	 * Determines whether the game should update the selection and cursor.
+	 * @param cap The current edit-tools capability.
+	 * @return True if the game has successfully begun dragging.
+	 */
+	public static boolean shouldUpdateDrag(IEditTools cap) { return cap.getEditPos1() != null; }
```

Comment:
`cap.isEditDragging()` doesn't update until the end of the tick, so we need to use `cap.getEditPos1()` in order to make sure the selection updates on the same tick that it starts, to prevent latency, and also to prevent weird bugs with the left-click and right-click event listeners.

Reply:
Why not have `isEditDragging()` update when dragging starts then?

Reply:
Oh wow you're right. For some reason I thought that would break something in the packet, but it works fine actually.